File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/data/EventDispatcher.java
package org.kevoree.modeling.api.data;

import org.kevoree.modeling.api.KEvent;
import org.kevoree.modeling.api.ModelListener;
import org.kevoree.modeling.api.event.DefaultKEvent;

import java.util.HashMap;

/**
 * Created by duke on 10/8/14.
 */
public class EventDispatcher {

    private HashMap<ModelListener, TimedRegistration> listeners = new HashMap<ModelListener, TimedRegistration>();

    public void register(ModelListener listener, long from, long to, String pathRegex) {
        listeners.put(listener, new TimedRegistration(from, to, pathRegex));
    }

    public void unregister(ModelListener listener) {
        listeners.remove(listener);
    }

    public void dispatch(DefaultKEvent event) {
        for (ModelListener l : listeners.keySet()) {
            TimedRegistration registration = listeners.get(l);
            if (registration.covered(event)) {
                l.on(event);
            }
        }
    }

    public void clear() {
        listeners.clear();
    }


    private class TimedRegistration {

        private Long from;
        private Long to;
        private String pathRegex;

        private TimedRegistration(Long from, Long to, String pathRegex) {
            this.from = from;
            this.to = to;
            this.pathRegex = pathRegex;
        }

        public boolean covered(KEvent event) {
            if (from != null) {
                if (from < event.src().now()) {
                    return false;
                }
            }
            if (to != null) {
                if (to < event.src().now()) {
                    return false;
                }
            }
            /*
            if (event.getSource() != null) {
                if (pathRegex.contains("*")) {
                    STRING regexPath = pathRegex.replace("*", ".*");
                    return event.getSource().uuid().matches(regexPath);
                } else {
                    return event.getSource().uuid().equals(pathRegex);
                }
            } else {
                return false;
            }*/

            return false;
        }

    }


}

File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/persistence/EventDispatcher.kt
package org.kevoree.modeling.api.persistence

import org.kevoree.modeling.api.events.ModelElementListener
import java.util.HashMap
import org.kevoree.modeling.api.events.ModelEvent
import org.kevoree.modeling.api.time.TimeAwareKMFContainer

/**
 * Created by duke on 8/22/14.
 */

class EventDispatcher {

    private var listeners = HashMap<ModelElementListener, TimedRegistration>()

    fun register(listener: ModelElementListener, from: Long?, to: Long?, pathRegex: String) {
        listeners.put(listener, TimedRegistration(from, to, pathRegex))
    }

    fun unregister(listener: ModelElementListener) {
        listeners.remove(listener)
    }

    fun dispatch(event: ModelEvent) {
        for (l in listeners) {
            if (l.value.covered(event)) {
                l.key.elementChanged(event)
            }
        }
    }

    fun clear() {
        listeners.clear()
    }

}

data class TimedRegistration(val from: Long?, val to: Long?, val pathRegex: String) {

    fun covered(event: ModelEvent): Boolean {
        if (event.source is TimeAwareKMFContainer<*>) {
            if (from != null) {
                if (from < event.source.now) {
                    return false
                }
            }
            if (to != null) {
                if (to < event.source.now) {
                    return false
                }
            }
        }
        if (event.source != null) {
            if (pathRegex.contains("*")) {
                var regexPath = pathRegex.replace("*", ".*")
                return event.source.path().matches(regexPath)
            } else {
                return event.source.path().equals(pathRegex)
            }
        } else {
            return false
        }
    }

}
